pandas/core/strings.py

```diff
@@ -309,20 +309,24 @@ def str_replace(arr, pat, repl, n=-1, case=True, flags=0):
     -------
     replaced : Series/Index of objects
     """
-    use_re = not case or len(pat) > 1 or flags
+    if is_string_like(repl):  # Check whether repl is valid (GH 13438)
+        use_re = not case or len(pat) > 1 or flags
 
-    if use_re:
-        if not case:
-            flags |= re.IGNORECASE
-        regex = re.compile(pat, flags=flags)
-        n = n if n >= 0 else 0
+        if use_re:
+            if not case:
+                flags |= re.IGNORECASE
+            regex = re.compile(pat, flags=flags)
+            n = n if n >= 0 else 0
 
-        def f(x):
-            return regex.sub(repl, x, count=n)
-    else:
-        f = lambda x: x.replace(pat, repl, n)
+            def f(x):
+                return regex.sub(repl, x, count=n)
+        else:
+            f = lambda x: x.replace(pat, repl, n)
 
-    return _na_map(f, arr)
+        return _na_map(f, arr)
+    else:
+        raise TypeError("Can't convert {0} object to str implicitly"
```

Comment:
Further, I don't find the error message clear. I know this is the one from the stdlib, but a numeric value can be converted to string, so that sounds a bit strange. Maybe just something like "expected a string" or "the replacement argument should be a string"